src/include/channel.h

```diff
@@ -276,7 +282,16 @@ class ShmChannel {
     int retries = 5;
     int ret;
     do {
-      ret = __machnet_channel_buf_free_bulk(ctx_, 1, index);
+      // clear and free buffer
+      MachnetMsgBuf_t *msg_buf = __machnet_channel_buf(ctx_, index[0]);
+      __machnet_channel_buf_init(msg_buf);
+      if (cached_buf_indices.size() < CACHED_BUF_SIZE) {
+        cached_buf_indices.push_back(index[0]);
+        cached_bufs.push_back(msg_buf);
+        ret = 1;
+      } else {
+        ret = __machnet_channel_buf_free_bulk(ctx_, 1, index);
```

Comment:
(The following comment should also be true for the bulk equivalent method.)

I think that this might not be the most efficient implementation of a cache.

On the `alloc` side it is clear, that you do a bulk allocation everytime the cache is empty. Here, however, you might have the phenomenon that the cache decays to single deallocations. Consider the following:

There is one flow that has `cwnd` worth of packets in flight. The remote side is going to send an `ACK` for every packet that it receives. Actually, it is highly likely that these packets will be received in batches. If your cache is full, you'd end up doing single `free` calls in the global pool (we release a buffer in the channel for each acknowledgment).

A more efficient implementation would consider operating on a certain watermark/stride say `BUFFER_CACHE_SIZE/2`. Then on `free`:
1. if the cache size is  less than `BUFFER_CACHE_SIZE`, you release to the cache and return.
2. if the cache size is equal to BUFFER_CACHE_SIZE, you release `BUFFER_CACHE_SIZE/2` to the global pool.